### PR TITLE
fix: look for associated types on parent traits and impls

### LINKED ISSUE #11562: Accessing associated types of parent trait fails in type signatures
### Aim

Extracted from #11545 to be a very specific case.

Relevant code:

```noir
trait KeyType {
    type Key;
}

trait Lookup: KeyType {
    fn lookup(self, key: Self::Key);
}

impl KeyType for u32 {
    type Key = Field;
}

impl Lookup for u32 {
    fn lookup(self, _key: Self::Key) {}
//                        ^^^^ Could not resolve 'Self' in path
}
```

In reality it fails because if tries to lookup a `Key` associated type for `Lookup` but it doesn't find one. The reason is that it's defined in the parent trait of `Lookup`, and this isn't currently done.

Similarly:

```noir
trait KeyType {
    type Key;
}

trait Lookup: KeyType {}

fn find<M: Lookup>(m: M, key: M::Key) {}
                              ^ Could not resolve 'M' in path
```

It fails in a different code path in the compiler but for the same reason: we aren't searching in the parent traits.

In both cases it works if `Key` is **also** defined in `Lookup`:

```noir
trait KeyType {
    type Key;
}

trait Lookup: KeyType {
    type Key;
}

fn find<M: Lookup>(m: M, key: M::Key) {}
```

**However**, Rust gives an error in this case saying that `M::Key` is ambiguous because it could either refer to `Lookup::Key` or `KeyType::Key` and it asks to disambiguate with `<M as Lookup>::Key` or `<M as KeyType>::Key`. Noir should likely behave in the same way.

### Expected Behavior

The two first snippets should compile, and the last one should error.

### Bug

The two first snippets don't compile, and the last one does.

### To Reproduce

1.
2.
3.
4.

### Workaround

None

### Workaround Description

_No response_

### Additional Context

_No response_

### Project Impact

None

### Blocker Context

_No response_

### Nargo Version

_No response_

### NoirJS Version

_No response_

### Proving Backend Tooling & Version

_No response_

### Would you like to submit a PR for this Issue?

None

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by guipublic**
# Description

## Problem

Resolves #11562

## Summary
Look for associated types in the parents hierarchy for traits and trait impls.
Trait constraints are added for both parent traits and trait impls.

## User Documentation

Check one:
- [X] No user documentation needed.
- [ ] Changes in _docs/_ included in this PR.
- [ ] **[For Experimental Features]** Changes in _docs/_ to be submitted in a separate PR.

# PR Checklist

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by asterite** [comment]
I'll review this now, but I noticed that (I think!) we don't have a test for when an associated type is defined in both the parent and the child:

```noir
trait KeyType {
    type Key;
}

trait Lookup: KeyType {
    type Key;
}

fn find<M: Lookup>(m: M, key: M::Key) {}
```

It works as expected, producing an ambiguity error, but I couldn't find such test.

**Comment by guipublic** [comment]
> I'll review this now, but I noticed that (I think!) we don't have a test for when an associated type is defined in both the parent and the child: (...)

I added the test.

File: compiler/noirc_frontend/src/elaborator/traits.rs [hunk 1]
```diff
@@ -589,6 +589,130 @@ impl Elaborator<'_> {
...
+                let named = vecmap(&instantiated.trait_generics.named, |named_type| {
+                    let fresh_id = self.interner.next_type_variable_id();
+                    let kind = named_type.typ.kind();
+                    let type_var = TypeVariable::unbound(fresh_id, kind);
+
+                    let assoc_type_id = parent_trait
+                        .associated_types
+                        .iter()
+                        .find(|a| a.name.as_ref() == named_type.name.as_str())
+                        .map_or(fresh_id, |a| a.type_var.id());
```
Comment (asterite) [inline compiler/noirc_frontend/src/elaborator/traits.rs]: I just tried it and it seems to work fine with `map` instead of `map_or`. We'd need to change the signature of `into_implicit_named_generic` to take an `Option`, but I think it makes more sense like that because if we can't find a parent associated type then there's no original type var to refer to.

File: compiler/noirc_frontend/src/tests/traits/trait_associated_items.rs [hunk 2]
```diff
@@ -1863,6 +1868,117 @@ fn associated_constant_in_return_type_with_generic_impl_forwarding() {
...
+        type Key;
+    }
+
+    trait Bar {
+        type Key;
+    }
+
+    fn ambiguous<M>(_m: M, _k: M::Key) where M: Foo + Bar {}
+       ^^^^^^^^^ unused function ambiguous
+       ~~~~~~~~~ unused function
```
Comment (asterite) [inline compiler/noirc_frontend/src/tests/traits/trait_associated_items.rs]: We could make this function public so we don't get the noise from these warnings. Same with other tests (we can use `_` for variable names, etc.)

File: compiler/noirc_frontend/src/elaborator/types.rs [hunk 3]
```diff
@@ -285,32 +286,129 @@ impl Elaborator<'_> {
...
+                return Some(typ);
             }
         }
+
+        None
+    }
+
+    /// Search for an associated type in parent 'trait impls'.
+    /// Used inside trait impls to resolve `Self::Foo` when `Foo` is defined on a parent trait.
+    /// For example, in `impl Lookup for u32`, resolving `Self::Key` where `Key` is from `KeyType`.
```
Comment (asterite) [inline compiler/noirc_frontend/src/elaborator/types.rs]: I think this comment refers to a particular test, and without knowing what test this is it's less clear. Maybe we can include a small snippet here?

File: compiler/noirc_frontend/src/elaborator/types.rs [hunk 4]
```diff
@@ -285,32 +286,129 @@ impl Elaborator<'_> {
     }
 
     /// Resolve `Self::Foo` to an associated type on the current trait or trait impl.
+    /// Also searches parent traits.
```
Comment (asterite) [inline compiler/noirc_frontend/src/elaborator/types.rs:289]: I think here we stop on the first match instead of gathering all matches and giving an ambiguity error. For example this code:

```noir
trait KeyType {
    type Key;
}

trait Lookup: KeyType {
    type Key;

    fn find(_: Self::Key) {}
}
```

gives an ambiguity error in Rust but currently none in this PR.

We could either fix it here or capture it as a follow-up issue.

File: compiler/noirc_frontend/src/elaborator/traits.rs (hunk 1 above)
Comment (asterite) [reply compiler/noirc_frontend/src/elaborator/traits.rs]: Also, running all frontend tests and all test_programs, in all cases `assoc_type_id` was never `None`. I don't know if it's a coincidence, but if not, maybe we could `expect` here instead of `map_or`.

File: compiler/noirc_frontend/src/elaborator/types.rs (hunk 3 above)
Comment (guipublic) [reply compiler/noirc_frontend/src/elaborator/types.rs]: Done, I removed the confusing comments

File: compiler/noirc_frontend/src/tests/traits/trait_associated_items.rs (hunk 2 above)
Comment (guipublic) [reply compiler/noirc_frontend/src/tests/traits/trait_associated_items.rs]: Done

File: compiler/noirc_frontend/src/elaborator/traits.rs (hunk 1 above)
Comment (guipublic) [reply compiler/noirc_frontend/src/elaborator/traits.rs]: I tried expect, let's see.

Comment (guipublic) [reply compiler/noirc_frontend/src/elaborator/traits.rs]: it works

File: compiler/noirc_frontend/src/elaborator/types.rs (hunk 4 above)
Comment (guipublic) [reply compiler/noirc_frontend/src/elaborator/types.rs:289]: Done, I added a check for this.
